### PR TITLE
ICU-22793 Clang-Tidy: google-readability-casting

### PR DESCRIPTION
**Pull request by roubert**
https://releases.llvm.org/17.0.1/tools/clang/tools/extra/docs/clang-tidy/checks/google/readability-casting.html

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22793
- [x] Required: The PR title must be prefixed with a JIRA Issue number.
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number.
- [x] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by eggrobin** [comment]
Having looked at all of icu4c/source/common, this certainly looks like an improvement, considering the static casts mixed with reinterpret casts that were lurking in various places, and the unnecessary casts.

Only 388 more files to look at…

**Comment by roubert** [comment]
On Thu, Jul 4, 2024 at 3:21 PM Robin Leroy ***@***.***> wrote:

> Only 388 more files to look at…

;-) See also my chat messages.

-- 
Fredrik Roubert
***@***.***

**Comment by eggrobin** [comment]
Done with icu4c/source/i18n/, only one comment so far.

Miscellaneous thoughts: turning the infamous Chinese calendar cast-with-comment « cast away const » into a `const_cast` is certainly a very good thing ([though I still think we should get rid of them eventually](https://github.com/unicode-org/icu/pull/2797#discussion_r1457265544)).
Even more so for turning casts without comments into `const_cast`s in icu4c/source/i18n/dtfmtsym.cpp.

263 more files to look at…

**Review by eggrobin — APPROVED** [review]

File: icu4c/source/i18n/csr2022.cpp [hunk 1]
```diff
@@ -46,7 +46,7 @@ int32_t CharsetRecog_2022::match_2022(const uint8_t *text, int32_t textLen, cons
             escN = 0;
             while(escN < escapeSequences_length) {
                 const uint8_t *seq = escapeSequences[escN];
-                int32_t seq_length = (int32_t)uprv_strlen((const char *) seq);
+                int32_t seq_length = static_cast<int32_t>(uprv_strlen((const char*)seq));
```
[the remark this answers is not in the document]
Comment (roubert) [reply icu4c/source/i18n/csr2022.cpp:49]: I don't have the faintest idea why the google-readability-casting check didn't fix the second one of those casts here, apart from the obvious answer that it's far from perfect. We'll need a round of manual fix-up for the rest.

**Comment by roubert** [comment]
> Miscellaneous thoughts: turning the infamous Chinese calendar cast-with-comment « cast away const » into a `const_cast` is certainly a very good thing ([though I still think we should get rid of them eventually](https://github.com/unicode-org/icu/pull/2797#discussion_r1457265544)).

I've already filed [ICU-22640](https://unicode-org.atlassian.net/browse/ICU-22640) about all the `const_cast<…>(this)`.

**Pull request merged**